Add prompt for credentials and conf file

facundovictor opened the pull request:
This PR adds support for:
* Using the configuration file located at `~/.config/cabinet/cli.ini`
* Setting a default vault name in config file.
* Setting an account in config file.
* Prompting for the account password

Also it adds the arguments for passing the `account` and a `vault` name. If they're passed, then they will be prioritized over the values set in the config file.

facundovictor committed:
Add prompt for credentials and config file support

Add support for:
- Prompt for password
- Passing the account by argument (-a, --account)
- Passing the vault name by argument (-v, --vault)
- Read the configuration file 'cli.ini', which contains the default
  vault name, and the account.

NOTE: The cli arguments have precedence over the configured arguments.

facundovictor committed:
Add README info for usig the command line

facundovictor committed:
Fix config file parsing

Add usage of section index for getting values from the config file

ivanalejandro0 commented on README.md:
````diff
@@ -1,3 +1,112 @@
+## Installation
+
+```
+pip install -r requiremen.txt
````
typo `s/requiremen/requirements/`

facundovictor replied:
fixed

ivanalejandro0 commented on README.md:
```diff
@@ -1,3 +1,112 @@
+
+By default, the cli will read the configuration file located on `~/.config/cabinet/cli.ini`. So far, it only the following parameters are configurable:
+
+* `account` : The users account.
```
*the user's account

facundovictor replied:
fixed

ivanalejandro0 commented on cli.py:
```diff
@@ -34,25 +36,51 @@ class CabinetWrapper:
+
+        if vault_name:
+            self.vault_name = vault_name
+        elif self.config['Credentials']['default_vault']:
```
This will raise an exception if there's no config file and you didn't specify a vault as a parameter, see:

```
➜ python cli.py search
Traceback (most recent call last):
  File "cli.py", line 302, in <module>
    app.run()
  File "/home/ivan/.virtualenvs/cabinet/lib/python3.5/site-packages/cement/core/foundation.py", line 882, in run
    return_val = self.controller._dispatch()
  File "/home/ivan/.virtualenvs/cabinet/lib/python3.5/site-packages/cement/core/controller.py", line 471, in _dispatch
    return func()
  File "/home/ivan/.virtualenvs/cabinet/lib/python3.5/site-packages/cement/core/controller.py", line 477, in _dispatch
    return func()
  File "cli.py", line 284, in default
    if cab.load_credentials(vault_name, account_id):
  File "cli.py", line 64, in load_credentials
    elif self.config['Credentials']['default_vault']:
  File "/usr/lib/python3.5/configparser.py", line 956, in __getitem__
    raise KeyError(key)
KeyError: 'Credentials'
```

You should

ivanalejandro0 replied:
oops, I forgot to finish that phrase... you get the idea though :)

I guess I was trying to say that you should safetly query for configs or something like that.

facundovictor replied:
Fixed

ivanalejandro0 commented on cli.py:
```diff
@@ -34,25 +36,51 @@ class CabinetWrapper:
+
+        if account_id:
+            self.account_id = account_id
+        elif self.config['Credentials']['account']:
```
same considerations as for vault config.

facundovictor replied:
Fixed

ivanalejandro0 commented on cli.py:
```diff
@@ -167,6 +190,10 @@ class Meta:
         arguments = [
             (['name'],
              dict(help='The item name.', action='store')),
+            (['-a', '--account'],
```
I think that both `--account` and `--vault` should be a parameter for the cli itself, not for its subcommands (`item` and `search`).
If you agree and we can do that easily please make the change... If not, we can leave it like this to get this out of the door asap.

facundovictor replied:
I set a `COMMON_ARGS` to include `--account` and `--vault` as global arguments. Cement doesn't provide a better approach. I took a look and It has a newest alternative to the `CementBaseController` that could help, but it changes everything, and I don't like it so far.

facundovictor committed:
Fix typos

facundovictor committed:
Add dictionary key existence check

Add key existence check for the dictionary attributes:
 - account
 - default_vault

facundovictor committed:
Set account and vault as parameters to the cli

Change the account and vault parameters to be available in all
the commands/subcommands.

ivanalejandro0 commented:
It seems that I can't use the parameters as arguments of the cli, see:
```
➜ python cli.py search -a ivan -v demo
Password: 
The following items was found:

➜ python cli.py -a ivan -v demo search
usage: cabinet (sub-commands ...) [options ...] {arguments ...}
cabinet: error: unrecognized arguments: search
```

Maybe I'm using it wrong?

What I suggested before was that the account/vault parameters are options for the cli, not for any of the two controllers (item/search). But, now that I've tested it a bit more, it may make sense to leave those options on the controllers so we can append the account/vault at the end if we want to. Is that your thinking?

facundovictor commented:
So far, **Cement** doesn't provide a way to add arguments to the cli before passing a sub-command. As the help is displaying:
`usage: cabinet (sub-commands ...) [options ...] {arguments ...}`

You can see that `search` is a sub-command and has to be present before the arguments.

I think that Cement is limiting us now, and it changes drastically if we use `ArgparseController` (the new version of the `CementBaseController`). We may have to pick another framework in the future.
